[CAP] Abstraction of actor_connector to go along with runtime factory and runtime abstraction

rajan-chari opened the pull request:
<!-- Thank you for your contribution! Please review https://microsoft.github.io/autogen/docs/Contribute before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

AutoGen CAP comes with a default runtime, that's ready to use out of the box, to make it easy to start building Agents in minutes.  However, different use cases and environments may require their own runtimes to host AutoGen CAP Agents.  Runtime factories were added to make it easy to create new runtimes.  

Along with runtime factories runtimes were also abstracted so that Agents/Actors can switch to different runtime implementations as needed.  

This PR introduces and abstraction for agent connectors since they will depend on the needs of each runtime.  Default implementation is also provided to make it super easy to start building Agents with AutoGen CAP.

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

thinkall commented on samples/apps/cap/py/autogencap/ag_adapter/AG2CAP.py:
```diff
@@ -4,7 +4,7 @@
+from .autogen_connector import AutoGenConnector
 
 
 class AG2CAP(ConversableAgent):
```
Would you like to also rename this file to something like "ag2cap.py"?

rajanjedi replied:
Yes.  Makes sense.  Thanks.

rajan-chari replied:
I am reconsidering some abstractions for the actor itself.  Converting this to draft.  Thanks @thinkall

thinkall reviewed (commented):
Thanks for the PR, @rajan-chari . I saw you renamed some filenames from upper cases to lower cases, would you like to rename others?

heyitsaamir commented on samples/apps/cap/py/autogencap/actor_connector.py:
```diff
@@ -0,0 +1,43 @@
+from abc import ABC, abstractmethod
+
+
+class IActorSender(ABC):
```
It would be good to document each method and intention here

rajan-chari replied:
Yeah good point.  Thanks.

rysweet commented:
This PR is against AutoGen 0.2. AutoGen 0.2 has been moved to the 0.2 branch. Please rebase your PR on the 0.2 branch or update it to work with the new AutoGen 0.4 that is now in main.

rysweet commented:
Hi Rajan,
we've rebased and updated the branch for you. If you are still considering bringing this PR forward please go ahead and update and mark as ready for review. You may also consider having a closer look at 0.4 which has moved to an event-driven actor model.  If you are not going to finish this PR please go ahed and close it.

rajan-chari commented:
> Hi Rajan, we've rebased and updated the branch for you. If you are still considering bringing this PR forward please go ahead and update and mark as ready for review. You may also consider having a closer look at 0.4 which has moved to an event-driven actor model. If you are not going to finish this PR please go ahed and close it.

Hi Ryan,
I will update the PR and mark it as ready for review in a short while.
